### DISCUSSION TITLE
TimestampParserError: unhandled timestamp format

### DISCUSSION THREAD

**Discussion by chriswardo** [aws/aws-sdk-js#4081] [General]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js) for answers
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js/issues) and didn't find any solution
  
**Describe the bug**
Using s3's `getObject` or `headObject` functions on a valid file, it gives the error `"TimestampParserError: unhandled timestamp format: un, 04 Oct 2020 12:29:09 GMT"`. _Note that "un" isn't a typo and is  the reason for the error judging by the source code of parseTimestamp in the SDK. The first character is somehow missing from the rfc822 date format._

The error happens only on iOS in a Cordova app. The exact same code (and same file) works in browsers (in a React project) and the Cordova app on Android. It even works in Safari on the same iPhone.

I've managed to get it working by adding an additional case to the `parseTimestamp` function in util.js:

```
else if (value.match(/^\w{2},/)) { // Broken rfc822
  return new Date(value.substr(4));
}
```

**Is the issue in the browser/Node.js?**
Node.js

**If on Node.js, are you running this on AWS Lambda?**
No

**Details of the browser/Node.js version**
v14.2.0

**SDK version number**
2.766.0

**Comment by chriswardo** [comment]
I forgot to mention that the device where the bug is happening is an iPhone 5S running iOS 12.4.8 so it's old but that doesn't explain the strangeness of the bug where the date has the first character missing, causing an error that means it can't get objects from S3. Thanks!

**Comment by ajredniwja** [comment]
Hey @chrisradek thanks for opening this issue, I am a little confused, how is the `un` getting added? If you could point me towards that.

**Comment by ajredniwja** [comment]
I would be happy to take or make a PR once I understand the case better.

**Comment by chriswardo** [comment]
Hi @ajredniwja the `un` is "Sun" (Sunday) but with the first character missing, presumably from the server somehow. It misses the first character no matter what day of the week it is. My temporary fix above looks for a date format with only 2 characters instead of the 3 that's already looked for to detect rfc822 in util.js

Note that I'm not doing anything abnormal. I'm simply using s3's `getObject` and `headObject` functions on a valid file. Those functions fail on the one device with the error I specified in my first post. I'm not doing anything with dates, this is purely an error from the SDK, which is why it's such a mystery why the first character is missing.

Please let me know if you need more clarification. Thanks!

**Comment by ajredniwja** [comment]
@chrisradek, thanks for the clarification, my only concern that we should actually look into why there is a missing character there, since it works well with the browser and Cordova app on Android. I have looked at your code which seems to be the right solution for your use case. The correct fix would be to fix the root case and not customize the SDK according to an use case.

**Comment by chriswardo** [comment]
@ajredniwja Agreed!

**Comment by ajredniwja** [comment]
Since it is not an issue with the SDK, I would like to close the issue, please reach out if you have any other questions.

**Comment by chriswardo** [comment]
@ajredniwja Are you sure it's not? All I know is that the first character is being removed at some point. It's either an issue with AWS server side or it's somewhere in the SDK. It needs to be investigated rather than assumed to not be an issue with the SDK imo.

**Comment by chriswardo** [comment]
@ajredniwja The exact same bug just happened to one of my users so it's definitely not limited to just my test device (not that I ever thought it was).

Can someone from AWS please investigate?

**Comment by ajredniwja** [comment]
I can investigate this further but it wasn't my assumption that the SDK is not at fault here. I did run some tests, I can share my findings once I redo them.

**Comment by chriswardo** [comment]
@ajredniwja Just to let you know, the user experienced the bug on an iPad Mini but unfortunately they weren't able to tell me what version of iOS it happened on as they upgraded between seeing the bug and me asking.

Happy new year :)

**Comment by fiveironfrnzy08** [comment]
I just came across this same issue and, using the same added else case, the issue was fixed. This is happening on iOS Cordova as well, with an iPhone 12 mini on 14.7.1.
`"TimestampParserError: unhandled timestamp format: at, 17 Apr 2021 14:13:29 GMT"`

I only noticed this come up after adding this cordova plugin. @chriswardo, do you know if you're using this as well?
https://github.com/globules-io/cordova-plugin-ios-xhr

**Comment by chriswardo** [comment]
@fiveironfrnzy08 Sorry I've only just noticed your comment. I'm not using that plugin, no.

I'm glad someone else is experiencing the same bug, so it doesn't look like I'm mad ;)

**Comment by akasummer** [comment]
I've created an AWS Lambda function to convert images to webp format (Triggered by CloudFront) and I am getting same error:
```
TimestampParserError: unhandled timestamp format: 01/01/2031
    at Object.parseTimestamp (/var/runtime/node_modules/aws-sdk/lib/util.js:346:11)
    at TimestampShape.toType (/var/runtime/node_modules/aws-sdk/lib/model/shape.js:309:22)
    at Object.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/protocol/rest.js:130:29)
    at Object.each (/var/runtime/node_modules/aws-sdk/lib/util.js:512:32)
    at Object.extractData (/var/runtime/node_modules/aws-sdk/lib/protocol/rest.js:113:8)
    at Request.extractData (/var/runtime/node_modules/aws-sdk/lib/protocol/rest_xml.js:66:8)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:688:14)
    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)
```

Piece of the code which is triggering this error:
```
S3.getObject({ Bucket: BUCKET, Key: originalKey }).promise()
      // perform the conversion
      .then(data => Sharp(data.Body)
        .webp({ quality: 80 })
        .toBuffer()
      )
```

After investigation, I've found out that the issue was that the object in S3 (the one I was trying to fetch), had a System-defined metadata: `Expires: 01/01/2031`. So, whole `getObject()` function was failing because of a parsing error of it's metadata value.
